### PR TITLE
New Feature: Enhanced Progress Control

### LINKED ISSUE #1626: Add dynamic label to progress visual element
### Description

Add a dynamic label property to the progress visual element. More than the value is needed to understand the progress of a task/function/... Users should be able to use a label with the progress viz element; it would appear next to the element.

### Acceptance Criteria

- [ ] Ensure new code is unit tested, and check code coverage is at least 90%.
- [ ] Propagate any change on the demos and run all of them to ensure there is no breaking change.
- [ ] Ensure any change is well documented.

### Code of Conduct

- [X] I have checked the [existing issues](https://github.com/Avaiga/taipy/issues?q=is%3Aissue+).
- [ ] I am willing to work on this issue (optional)

### PR DESCRIPTION
**Pull request by namnguyen20999**
### New Feature: Enhanced Progress Control

This PR introduces several enhancements to the progress control element:

1. **Dynamic `title` Property:**
   - A new `title` property has been added to the progress control, allowing users to provide a dynamic title that can update based on the current status or context of the progress.

2. **Dynamic `title_anchor` Property:**
   - Introduced the `title_anchor` property, which determines the placement of the title around the progress element. Available options include `"left"`, `"right"`, `"top"`, `"bottom"`, or `"none"`.
   - When set to `"none"`, the title will be hidden. The default value is `"bottom"`.

**main.py**
```
from taipy.gui import Gui

value = 50
title = "Progress"
title_anchor = "right"  # Options: "left", "right", "top", "bottom", "none"

page = """
<|progress|value={value}|linear|title={title}|title_anchor={title_anchor}|show_value|>
"""

if __name__ == '__main__':
    Gui(page).run(use_reloader=True, debug=True")
```

**Result**
image

**Other cases:**
`title_anchor = "top" `
image

`title_anchor = "bottom" `
image

`title_anchor = "left" `
image

`title_anchor = "none" `
image

### CODE REVIEW

**Review by FlorianJacta — DISMISSED** [review]

File: taipy/gui/viselements.json [hunk 1]
```diff
@@ -1341,6 +1341,11 @@
                         "doc": "If set, then the value represents the progress percentage that is shown.TODO - if unset?",
                         "default_property": true
                     },
+                    {
```
Comment (FabienLelaquais) [inline taipy/gui/viselements.json]: Please move that property *after* 'title'. Kind of less important in my opitnion.

File: taipy/gui/viselements.json [hunk 2]
```diff
@@ -1353,6 +1358,17 @@
                         "default_value": "False",
                         "doc": "If set to True, the progress value is shown."
                     },
+                    {
+                        "name": "title",
+                        "type": "dynamic(str)",
+                        "doc": "The title of the progress indicator."
```
Comment (FabienLelaquais) [inline taipy/gui/viselements.json:1359]: Please check whether we use 'the <control>' or 'this <control>' terminology in the rest of the json file.
Just be homogeneous.

File: taipy/gui/viselements.json [hunk 3]
```diff
@@ -1341,6 +1341,11 @@
                         "doc": "If set, then the value represents the progress percentage that is shown.TODO - if unset?",
                         "default_property": true
                     },
+                    {
+                        "name": "color",
+                        "type": "str",
+                        "doc": "The color of the progress indicator."
```
Comment (FabienLelaquais) [inline taipy/gui/viselements.json]: Please check whether we use 'the ' or 'this ' terminology in the rest of the json file.
Just be homogeneous.

File: frontend/taipy-gui/src/components/Taipy/Progress.spec.tsx [hunk 4]
```diff
@@ -48,4 +47,94 @@ describe("Progress component", () => {
...
+    it('should return an empty string when title is falsy and titleAnchor is truthy', () => {
+        const result = getBoxWidth(undefined, "Anchor");
+        expect(result).toBe("");
+    });
+
+    it('should return an empty string when both title and titleAnchor are falsy', () => {
+        const result = getBoxWidth(undefined, undefined);
+        expect(result).toBe("");
+    });
+});
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/src/components/Taipy/Progress.spec.tsx]: Missing carriage return

**Review by FabienLelaquais — DISMISSED** [review]
👌

File: taipy/gui/viselements.json (hunk 2 above)
Comment (namnguyen20999) [reply taipy/gui/viselements.json:1359]: resolved

File: taipy/gui/viselements.json (hunk 1 above)
Comment (namnguyen20999) [reply taipy/gui/viselements.json]: resolved

File: taipy/gui/viselements.json (hunk 3 above)
Comment (namnguyen20999) [reply taipy/gui/viselements.json]: resolved

File: frontend/taipy-gui/src/components/Taipy/Progress.spec.tsx (hunk 4 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Progress.spec.tsx]: I've refactored the file to match CRLF

File: frontend/taipy-gui/src/components/Taipy/Progress.spec.tsx [hunk 5]
```diff
@@ -12,40 +12,144 @@
...
+
     it("renders linear progress with value (determinate)", () => {
         const { getByRole, getByText } = render(<Progress linear showValue value={50} />);
         const elt = getByRole("progressbar");
         const valueText = getByText("50%");
         expect(elt).toHaveClass("MuiLinearProgress-root");
         expect(valueText).toBeInTheDocument();
     });
+
+    it("should return null when render is false", async () => {
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/src/components/Taipy/Progress.spec.tsx]: The test label is not really relevant here.
The point would be "Should not render when...".
This 'return null' really is a technical aspect of the implementation.

File: taipy/gui/viselements.json [hunk 6]
```diff
@@ -1353,6 +1353,22 @@
...
+                        "name": "title",
+                        "type": "dynamic(str)",
+                        "doc": "The title of the progress indicator."
+                    },
+                    {
+                        "name": "color",
+                        "type": "str",
+                        "doc": "The color of the progress indicator."
+                    },
+                    {
```
Comment (FabienLelaquais) [inline taipy/gui/viselements.json]: Please move 'title_anchor' next (just after) 'title': they are so related.
My prefered order would be title, title_anchor, color.

**Review by FabienLelaquais — DISMISSED** [review]
👍

File: frontend/taipy-gui/src/components/Taipy/Progress.tsx [hunk 7]
```diff
@@ -45,51 +49,136 @@ const circularPrgSx = {
...
+    switch (titleAnchor) {
+        case "top":
+            return "column";
+        case "bottom":
+            return "column-reverse";
+        case "left":
+            return "row";
+        case "right":
+            return "row-reverse";
+        default:
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Progress.tsx:61]: I thought default was "bottom" ?

File: frontend/taipy-gui/src/components/Taipy/Progress.tsx [hunk 8]
```diff
@@ -45,51 +49,136 @@ const circularPrgSx = {
...
 
     return showValue && value !== undefined ? (
         linear ? (
-            <Box sx={linearSx} className={className} id={props.id}>
-                <Box sx={linearPrgSx}>
-                    <LinearProgress variant="determinate" value={value} />
-                </Box>
-                <Box sx={linearTxtSx}>
-                    <Typography variant="body2" color="text.secondary">{`${Math.round(value)}%`}</Typography>
+            <Box sx={{ ...linearSx, flexDirection: getFlexDirection(titleAnchor) }}>
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Progress.tsx]: sx value should be in a useMemo

File: frontend/taipy-gui/src/components/Taipy/Progress.tsx [hunk 9]
```diff
@@ -45,51 +49,136 @@ const circularPrgSx = {
...
+                {title && titleAnchor !== "none" ? (
+                    <Typography sx={{ margin: 1 }} variant="caption">
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Progress.tsx]: sx => const outside the component

File: frontend/taipy-gui/src/components/Taipy/Progress.tsx [hunk 10]
```diff
@@ -45,51 +49,136 @@ const circularPrgSx = {
...
+                        {title}
+                    </Typography>
+                ) : null}
+                <Box sx={{ ...linearSx, width: getBoxWidth(title, titleAnchor) }} className={className} id={props.id}>
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Progress.tsx]: useMemo

File: frontend/taipy-gui/src/components/Taipy/Progress.tsx [hunk 11]
```diff
@@ -45,51 +49,136 @@ const circularPrgSx = {
...
+                    <Box sx={linearPrgSx}>
+                        <LinearProgress
+                            sx={{
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Progress.tsx]: use Memo

File: frontend/taipy-gui/src/components/Taipy/Progress.tsx [hunk 12]
```diff
@@ -45,51 +49,136 @@ const circularPrgSx = {
...
                 </Box>
             </Box>
         ) : (
-            <Box sx={circularSx} className={className} id={props.id}>
-                <CircularProgress variant="determinate" value={value} />
-                <Box sx={circularPrgSx}>
-                    <Typography variant="caption" component="div" color="text.secondary">
-                        {`${Math.round(value)}%`}
+            <Box
+                sx={{
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Progress.tsx]: useMemo

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]
limit re-render by makeing sx more stable

File: frontend/taipy-gui/src/components/Taipy/Progress.spec.tsx (hunk 5 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Progress.spec.tsx]: Changed to a non-technical name

File: taipy/gui/viselements.json (hunk 6 above)
Comment (namnguyen20999) [reply taipy/gui/viselements.json]: updated

File: frontend/taipy-gui/src/components/Taipy/Progress.tsx (hunk 8 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Progress.tsx]: updated

File: frontend/taipy-gui/src/components/Taipy/Progress.tsx [hunk 13]
```diff
@@ -46,50 +50,129 @@ const circularPrgSx = {
...
-    const { linear = false, showValue = false } = props;
+    const { linear = false, showValue = false, titleAnchor = "bottom" } = props;
 
     const className = useClassNames(props.libClassName, props.dynamicClassName, props.className);
     const value = useDynamicProperty(props.value, props.defaultValue, undefined, "number", true);
     const render = useDynamicProperty(props.render, props.defaultRender, true);
+    const title = useDynamicProperty(props.title, props.defaultTitle, undefined);
+
+    const memoizedValues = useMemo(() => {
+        const getFlexDirection = (titleAnchor: string) => {
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Progress.tsx]: this could be defined outside the comp

**Review by FredLL-Avaiga — DISMISSED** [review]

File: frontend/taipy-gui/src/components/Taipy/Progress.tsx (hunk 13 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Progress.tsx]: resolved

**Review by FredLL-Avaiga — APPROVED** [review]

**Pull request merged**